### PR TITLE
Make number of CPUs configurable

### PR DESCRIPTION
**Pull request by magicmark**
This branch adds a `cpu` config option, to cap the number of workers Hypernova spawns

(This is desirable in deployments where we'd like to control the number of cpus per container)

Thanks!

### CODE REVIEW

File: src/coordinator.js [hunk 1]
```diff
@@ -5,7 +5,19 @@ import './environment';
...
+
+  if (cpus === 0) {
+    return realCount - 1 || 1;
+  }
+
+  if (cpus > realCount) {
+    logger.info(`Specified CPU count of ${cpus} is bigger than OS reported count of ${realCount}`);
+  }
+
+  return cpus - 1;
```
Comment (chriskuehl) [inline src/coordinator.js]: This seems fishy to me -- I request one CPU and end up with zero workers? Is that right?

File: README.md [hunk 2]
```diff
@@ -217,6 +217,8 @@ Options, and their defaults
   bodyParser: {
     limit: 1024 * 1000,
   },
+  // number of cpus hypernova will use. 0 to use all available
+  cpus: 0,
```
Comment (ljharb) [inline README.md]: I think it is very confusing to have zero mean "all" - I'd expect Infinity to mean "all".

File: src/coordinator.js (hunk 1 above)
Comment (ljharb) [inline src/coordinator.js]: I think making the "-1" behavior implicit combined with s configurable CPU count is problematic.

What if "cpus" was a function that took `realCount` and returned a number? The default would do the -1 logic; hypernova would throw if it returned a non-number; and it'd be easy to do `() => 3` or `X => X - 1`

**Comment by magicmark** [comment]
Thanks, good points, I will integrate feedback.

The -1 thing exists to mirror the existing behaviour - presumably the
master process that isn't a worker counts toward the count? i.e. we want
n-1 workers? Or is this not right?

File: src/coordinator.js (hunk 1 above)
Comment (chriskuehl) [inline src/coordinator.js]: Out of curiosity, why do you do the -1 behavior anyway? Is that an attempt to leave one CPU free for the main thread or something?

Comment (ljharb) [inline src/coordinator.js]: Yes, it's generally a bad practice to make a process take up 100% of the available CPUs; you always want one core free for `ps` and `kill -9`.

File: README.md [hunk 3]
```diff
@@ -226,8 +226,10 @@ Options, and their defaults
   // the port the app will start on
   port: 8080,
   // whether or not to run in parallel using all available cpus
-  endpoint: '/batch'
+  endpoint: '/batch',
   // default endpoint path
+  // return the number of cpus hypernova will use
+  getCPUs: undefined
```
Comment (ljharb) [inline README.md]: this should indicate that if null or undefined is passed, that the default will be `cpus => cpus - 1`

File: src/coordinator.js [hunk 4]
```diff
@@ -5,7 +5,21 @@ import './environment';
 import logger from './utils/logger';
 import { raceTo } from './utils/lifecycle';
 
-const WORKER_COUNT = os.cpus().length - 1 || 1;
+export function getWorkerCount(getCPUs) {
+  const realCount = os.cpus().length;
+
+  if (!getCPUs) {
```
Comment (ljharb) [inline src/coordinator.js]: i think instead we should define a function `getDefaultCPUs` in the module scope, and do `getCPUs = getDefaultCPUs` in the function signature.

Comment (ljharb) [inline src/coordinator.js]: Separately, if `typeof getCPUs !== 'function'`, we should throw a `TypeError`.

File: test/coordinator-test.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { assert } from 'chai';
+import os from 'os';
+import sinon from 'sinon';
+import { getWorkerCount } from '../lib/coordinator';
+
+describe('coordinator', () => {
+  it('uses the correct number of cpus', () => {
+    const sandbox = sinon.sandbox.create();
+    sandbox.stub(os, 'cpus').returns([{}, {}, {}, {}, {}]);
```
Comment (ljharb) [inline test/coordinator-test.js]: it'd be nice for the test if this array of 5 objects was stored in a variable, so that its length could be used in place of a hardcoded `4`

Comment (magicmark) [inline test/coordinator-test.js]: Hmm I generally try and stick to the 'no logic in tests' thing as much as possible - does asserting against the length of an object possibly violate this?

File: README.md (hunk 3 above)
Comment (magicmark) [inline README.md]: Let me know if the new text looks good (unsure exactly how much detail you're looking for here since I documented it further down in the readme. maybe we can kill that though if it's fully documented here instead?)

File: test/coordinator-test.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  it('default method returns correct number of cpus', () => {
+    assert.equal(getDefaultCPUs(5), 4, 'getDefaultCPUs returns n - 1 CPUs');
+    assert.equal(getDefaultCPUs(), 1, 'getDefaultCPUs falls back to 1 CPU');
+  });
+
+  it('uses the correct number of cpus', () => {
+    const sandbox = sinon.sandbox.create();
+    sandbox.stub(os, 'cpus').returns([{}, {}, {}, {}, {}]);
+
+    assert.equal(getWorkerCount(), 4, 'getWorkerCount defaults to all available cpus minus one');
```
Comment (magicmark) [inline test/coordinator-test.js]: I haven't mocked get `getDefaultCPUs` out for the sake of test readability, but lmk if you think it should be.

File: test/coordinator-test.js (hunk 5 above)
Comment (ljharb) [inline test/coordinator-test.js]: I think that a more important principle is that the relationship between values should be expressed in code, and not be implicit, even in tests.

File: src/coordinator.js [hunk 7]
```diff
@@ -5,7 +5,25 @@ import './environment';
...
+export function getWorkerCount(getCPUs = getDefaultCPUs) {
+  const realCount = os.cpus().length;
+
+  if (typeof getCPUs !== 'function') {
+    throw new TypeError('getCPUs must be a function');
+  }
+
+  const requested = getCPUs(realCount);
+
+  if (!Number.isInteger(requested)) {
```
Comment (ljharb) [inline src/coordinator.js]: We should probably also enforce that it's `> 0`

File: test/coordinator-test.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { assert } from 'chai';
+import os from 'os';
+import sinon from 'sinon';
+import { getDefaultCPUs, getWorkerCount } from '../lib/coordinator';
+
+describe('coordinator', () => {
+  it('default method returns correct number of cpus', () => {
+    assert.equal(getDefaultCPUs(5), 4, 'getDefaultCPUs returns n - 1 CPUs');
+    assert.equal(getDefaultCPUs(), 1, 'getDefaultCPUs falls back to 1 CPU');
```
Comment (ljharb) [inline test/coordinator-test.js]: Since this is exported, let's add a test, just in case, for what happens when 0 or a negative number is passed in (I'd expect it to throw or return `NaN`)

Comment (magicmark) [inline test/coordinator-test.js]: (fwiw I only exported it for the sake of testing. i guess we could scope it inside of `getWorkerCount` to avoid this complexity, but i'm not a fan of this approach. lmk if you'd prefer this.)

File: test/coordinator-test.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import { assert } from 'chai';
+import os from 'os';
+import sinon from 'sinon';
+import { getDefaultCPUs, getWorkerCount } from '../lib/coordinator';
+
+describe('coordinator', () => {
+  it('default method returns correct number of cpus', () => {
+    assert.equal(getDefaultCPUs(5), 4, 'getDefaultCPUs returns n - 1 CPUs');
+
+    assert.throws(getDefaultCPUs, TypeError, 'getDefaultCPUs must accept a positive integer');
```
Comment (magicmark) [inline test/coordinator-test.js]: These grouped `assert.throws` tests could be refactored to something like

```
[
    () => { getDefaultCPUs('three'); },
    () => { getDefaultCPUs(0); },
    getDefaultCPUs,
].forEach((tryGetDefault) => {
    assert.throws(tryGetDefault, TypeError, 'getDefaultCPUs must accept a positive integer');
});
```

not sure which is better overall for test readability/clarity, maybe doesn't matter too much

File: test/coordinator-test.js (hunk 8 above)
Comment (magicmark) [inline test/coordinator-test.js]: I've added some more tests and erred on the side of caution, hopefully looks ok now :)

File: README.md [hunk 10]
```diff
@@ -266,6 +269,12 @@ hypernova({
 });
 ```
 
+#### `getCPUs`
+
+This lets you specify the number of cores Hypernova will run workers on. Receives an argument containing the number of cores as reported by the OS.
+
+If this method is not overriden, or if a falsey value is passed, the default method will return the number of reported cores minus 1.
```
Comment (ljharb) [inline README.md]: `overriden → overridden`, `falsey → falsy`

File: src/coordinator.js [hunk 11]
```diff
@@ -5,7 +5,29 @@ import './environment';
 import logger from './utils/logger';
 import { raceTo } from './utils/lifecycle';
 
-const WORKER_COUNT = os.cpus().length - 1 || 1;
+export function getDefaultCPUs(realCount) {
+  if (!Number.isInteger(realCount) || (realCount <= 0)) {
```
Comment (ljharb) [inline src/coordinator.js]: i don't think the parens around the `<=` are needed.

File: README.md (hunk 10 above)
Comment (magicmark) [inline README.md]: derp, thanks

**Comment by ljharb** [comment]
@magicmark thanks! I'm going to wait for @goatslacker to review this, and it'd be good in the meantime to rebase this (to remove "typo fix" style commits, and all merge commits), but otherwise this is great :-)

**Comment by magicmark** [comment]
(Rebased to make this one commit. Pushed with `git push origin HEAD -f`,  had to resolve a whitespace merge conflict in github.)

**Comment by magicmark** [comment]
Rebased to make one commit.

**Comment by dadowns** [comment]
Looks like you need another rebase. I was going to write this last week and then noticed your PR, nice! I am very eager to see this released as I need to cap max workers before getting to prod.

**Comment by magicmark** [comment]
Pinging @goatslacker for review :)

**Comment by ljharb** [comment]
@magicmark please rebase on the command line; the "update" button on the github UI creates a merge commit.

**Comment by magicmark** [comment]
@ljharb @goatslacker Rebased again to make one commit :)

**Comment by Will-Sommers** [comment]
Hey y'all! Thank you for this one, we've been following this one for a while at Rent the Runway.

**Comment by goatslacker** [comment]
I'll put out a release and fix the changelog in a minute. I'd like to merge some of the other PRs as well.
